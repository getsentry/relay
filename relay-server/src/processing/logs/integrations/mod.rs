use relay_event_schema::protocol::{OurLog, OurLogHeader};
use relay_quotas::DataCategory;
use zstd::zstd_safe::WriteBuf;

use crate::envelope::{ContainerItems, Item, WithHeader};
use crate::integrations::{Integration, LogsIntegration};
use crate::managed::RecordKeeper;

mod otel;

/// Expands a list of [`Integration`] items into `result`.
///
/// The function expects *only* log item integrations.
pub fn expand_into(
    result: &mut ContainerItems<OurLog>,
    records: &mut RecordKeeper<'_>,
    items: Vec<Item>,
) {
    for item in items {
        let integration = match item.integration() {
            Some(Integration::Logs(integration)) => integration,
            integration => {
                records.internal_error(InvalidIntegration(integration), item);
                continue;
            }
        };

        let produce = |log: OurLog| {
            let byte_size = relay_ourlogs::calculate_size(&log);

            records.modify_by(DataCategory::LogItem, 1);
            records.modify_by(DataCategory::LogByte, byte_size as isize);

            result.push(WithHeader {
                header: Some(OurLogHeader {
                    byte_size: Some(byte_size),
                    other: Default::default(),
                }),
                value: log.into(),
            });
        };

        let payload = item.payload();
        let payload = payload.as_slice();

        let result = match integration {
            LogsIntegration::OtelV1 { format } => otel::expand(format, payload, produce),
        };

        match result {
            Err(err) => drop(records.reject_err(err, item)),
            Ok(()) => {
                // Undo all the base item quantities, as they will be completely taken over by the parsed
                // contents, which contains an arbitrary amount of items (even 0).
                for (category, quantity) in item.quantities() {
                    records.modify_by(category, -(quantity as isize));
                }
            }
        }
    }
}

#[derive(Debug, thiserror::Error)]
#[error("Expected a logs integration, got: {0:?}")]
struct InvalidIntegration(Option<Integration>);
