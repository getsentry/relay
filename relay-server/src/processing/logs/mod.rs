use std::sync::Arc;

use relay_event_schema::processor::ProcessingAction;
use relay_event_schema::protocol::OurLog;
use relay_filter::FilterStatKey;
use relay_pii::PiiConfigError;
use relay_quotas::{DataCategory, RateLimits};

use crate::Envelope;
use crate::envelope::{
    ContainerItems, ContainerWriteError, EnvelopeHeaders, Item, ItemContainer, ItemType, Items,
};
use crate::managed::{
    Counted, Managed, ManagedEnvelope, ManagedResult as _, OutcomeError, Quantities,
};
use crate::processing::{
    self, Context, Forward, Output, QuotaRateLimiter, RateLimited, RateLimiter, Rejected,
};
use crate::services::outcome::{DiscardReason, Outcome};

mod filter;
mod process;
#[cfg(feature = "processing")]
mod store;
mod utils;
mod validate;

pub use self::utils::get_calculated_byte_size;

type Result<T, E = Error> = std::result::Result<T, E>;

#[derive(Debug, thiserror::Error)]
pub enum Error {
    /// A duplicated item container for logs.
    #[error("duplicate log container")]
    DuplicateContainer,
    /// Logs filtered because of a missing feature flag.
    #[error("logs feature flag missing")]
    FilterFeatureFlag,
    /// Logs filtered due to a filtering rule.
    #[error("log filtered")]
    Filtered(FilterStatKey),
    /// The logs are rate limited.
    #[error("rate limited")]
    RateLimited(RateLimits),
    /// Internal error, Pii config could not be loaded.
    #[error("Pii configuration error")]
    PiiConfig(PiiConfigError),
    /// A processor failed to process the logs.
    #[error("envelope processor failed")]
    ProcessingFailed(#[from] ProcessingAction),
    /// The log is invalid.
    #[error("invalid: {0}")]
    Invalid(DiscardReason),
}

impl OutcomeError for Error {
    type Error = Self;

    fn consume(self) -> (Option<Outcome>, Self::Error) {
        let outcome = match &self {
            Self::DuplicateContainer => Some(Outcome::Invalid(DiscardReason::DuplicateItem)),
            Self::FilterFeatureFlag => None,
            Self::Filtered(f) => Some(Outcome::Filtered(f.clone())),
            Self::RateLimited(limits) => {
                let reason_code = limits.longest().and_then(|limit| limit.reason_code.clone());
                Some(Outcome::RateLimited(reason_code))
            }
            Self::PiiConfig(_) => Some(Outcome::Invalid(DiscardReason::ProjectStatePii)),
            Self::ProcessingFailed(_) => Some(Outcome::Invalid(DiscardReason::Internal)),
            Self::Invalid(reason) => Some(Outcome::Invalid(*reason)),
        };

        (outcome, self)
    }
}

/// A processor for Logs.
///
/// It processes items of type: [`ItemType::OtelLog`] and [`ItemType::Log`].
#[derive(Debug)]
pub struct LogsProcessor {
    limiter: Arc<QuotaRateLimiter>,
}

impl LogsProcessor {
    /// Creates a new [`Self`].
    pub fn new(limiter: Arc<QuotaRateLimiter>) -> Self {
        Self { limiter }
    }
}

impl processing::Processor for LogsProcessor {
    type UnitOfWork = SerializedLogs;
    type Output = LogOutput;
    type Error = Error;

    fn prepare_envelope(
        &self,
        envelope: &mut ManagedEnvelope,
    ) -> Option<Managed<Self::UnitOfWork>> {
        let headers = envelope.envelope().headers().clone();

        let otel_logs = envelope
            .envelope_mut()
            .take_items_by(|item| matches!(*item.ty(), ItemType::OtelLog))
            .into_vec();
        let logs = envelope
            .envelope_mut()
            .take_items_by(|item| matches!(*item.ty(), ItemType::Log))
            .into_vec();

        let work = SerializedLogs {
            headers,
            otel_logs,
            logs,
        };
        Some(Managed::from_envelope(envelope, work))
    }

    async fn process(
        &self,
        mut logs: Managed<Self::UnitOfWork>,
        ctx: Context<'_>,
    ) -> Result<Output<Self::Output>, Rejected<Error>> {
        validate::container(&logs)?;

        if ctx.is_proxy() {
            // If running in proxy mode, just apply cached rate limits and forward without
            // processing.
            //
            // Static mode needs processing, as users can override project settings manually.
            self.limiter.enforce_quotas(&mut logs, ctx).await?;
            return Ok(Output::just(LogOutput::NotProcessed(logs)));
        }

        // Fast filters, which do not need expanded logs.
        filter::feature_flag(ctx).reject(&logs)?;

        let mut logs = process::expand(logs, ctx);
        process::normalize(&mut logs);
        filter::filter(&mut logs, ctx);
        process::scrub(&mut logs, ctx);

        self.limiter.enforce_quotas(&mut logs, ctx).await?;

        Ok(Output::just(LogOutput::Processed(logs)))
    }
}

/// Output produced by [`LogsProcessor`].
#[derive(Debug)]
pub enum LogOutput {
    NotProcessed(Managed<SerializedLogs>),
    Processed(Managed<ExpandedLogs>),
}

impl Forward for LogOutput {
    fn serialize_envelope(self) -> Result<Managed<Box<Envelope>>, Rejected<()>> {
        let logs = match self {
            Self::NotProcessed(logs) => logs,
            Self::Processed(logs) => logs.try_map(|logs, r| {
                r.lenient(DataCategory::LogByte);
                logs.serialize()
                    .map_err(drop)
                    .with_outcome(Outcome::Invalid(DiscardReason::Internal))
            })?,
        };

        Ok(logs.map(|logs, r| {
            r.lenient(DataCategory::LogByte);
            logs.serialize_envelope()
        }))
    }

    #[cfg(feature = "processing")]
    fn forward_store(
        self,
        s: &relay_system::Addr<crate::services::store::Store>,
    ) -> Result<(), Rejected<()>> {
        let logs = match self {
            LogOutput::NotProcessed(logs) => {
                return Err(logs.internal_error(
                    "logs must be processed before they can be forwarded to the store",
                ));
            }
            LogOutput::Processed(logs) => logs,
        };

        let scoping = logs.scoping();
        let received_at = logs.received_at();

        let (logs, retention) = logs.split_with_context(|logs| (logs.logs, logs.retention));
        let ctx = store::Context {
            scoping,
            received_at,
            retention,
        };

        for log in logs {
            if let Ok(log) = log.try_map(|log, _| store::convert(log, &ctx)) {
                s.send(log)
            };
        }

        Ok(())
    }
}

/// Logs in their serialized state, as transported in an envelope.
#[derive(Debug)]
pub struct SerializedLogs {
    /// Original envelope headers.
    headers: EnvelopeHeaders,

    /// OTel Logs are not sent in containers, an envelope is very likely to contain multiple OTel logs.
    otel_logs: Vec<Item>,
    /// Logs are sent in item containers, there is specified limit of a single container per
    /// envelope.
    ///
    /// But at this point this has not yet been validated.
    logs: Vec<Item>,
}

impl SerializedLogs {
    fn serialize_envelope(self) -> Box<Envelope> {
        let mut items = self.logs;
        items.extend(self.otel_logs);
        Envelope::from_parts(self.headers, Items::from_vec(items))
    }

    fn items(&self) -> impl Iterator<Item = &Item> {
        self.otel_logs.iter().chain(self.logs.iter())
    }

    /// Returns the total count of all logs contained.
    ///
    /// This contains all logical log items, not just envelope items and is safe
    /// to use for rate limiting.
    fn count(&self) -> usize {
        self.items()
            .map(|item| item.item_count().unwrap_or(1) as usize)
            .sum()
    }

    /// Returns the sum of bytes of all logs contained.
    fn bytes(&self) -> usize {
        self.items().map(|item| item.len()).sum()
    }
}

impl Counted for SerializedLogs {
    fn quantities(&self) -> Quantities {
        smallvec::smallvec![
            (DataCategory::LogItem, self.count()),
            (DataCategory::LogByte, self.bytes())
        ]
    }
}

impl RateLimited for Managed<SerializedLogs> {
    type Error = Error;

    async fn enforce<T>(
        &mut self,
        mut rate_limiter: T,
        _ctx: Context<'_>,
    ) -> Result<(), Rejected<Self::Error>>
    where
        T: RateLimiter,
    {
        let scoping = self.scoping();

        let items = rate_limiter
            .try_consume(scoping.item(DataCategory::LogItem), self.count())
            .await;
        let bytes = rate_limiter
            .try_consume(scoping.item(DataCategory::LogByte), self.bytes())
            .await;

        let limits = items.merge_with(bytes);
        if !limits.is_empty() {
            return Err(self.reject_err(Error::RateLimited(limits)));
        }

        Ok(())
    }
}

/// Logs which have been parsed and expanded from their serialized state.
#[derive(Debug)]
pub struct ExpandedLogs {
    /// Original envelope headers.
    headers: EnvelopeHeaders,
    /// Expanded and parsed logs.
    logs: ContainerItems<OurLog>,

    // These fields are currently necessary as we don't pass any project config context to the
    // store serialization. The plan is to get rid of them by giving the serialization context,
    // including the project info, where these are pulled from: #4878.
    /// Retention in days.
    #[cfg(feature = "processing")]
    retention: Option<u16>,
}

impl Counted for ExpandedLogs {
    fn quantities(&self) -> Quantities {
        smallvec::smallvec![
            (DataCategory::LogItem, self.logs.len()),
            (DataCategory::LogByte, self.bytes())
        ]
    }
}

impl ExpandedLogs {
    /// Returns the total count of all logs contained.
    fn count(&self) -> usize {
        self.logs.len()
    }

    /// Returns the sum of bytes of all logs contained.
    fn bytes(&self) -> usize {
        self.logs.iter().map(get_calculated_byte_size).sum()
    }

    fn serialize(self) -> Result<SerializedLogs, ContainerWriteError> {
        let mut logs = Vec::new();

        if !self.logs.is_empty() {
            let mut item = Item::new(ItemType::Log);
            ItemContainer::from(self.logs)
                .write_to(&mut item)
                .inspect_err(|err| relay_log::error!("failed to serialize logs: {err}"))?;
            logs.push(item);
        }

        Ok(SerializedLogs {
            headers: self.headers,
            otel_logs: Default::default(),
            logs,
        })
    }
}

impl RateLimited for Managed<ExpandedLogs> {
    type Error = Error;

    async fn enforce<T>(
        &mut self,
        mut rate_limiter: T,
        _ctx: Context<'_>,
    ) -> Result<(), Rejected<Self::Error>>
    where
        T: RateLimiter,
    {
        let scoping = self.scoping();

        let items = rate_limiter
            .try_consume(scoping.item(DataCategory::LogItem), self.count())
            .await;
        let bytes = rate_limiter
            .try_consume(scoping.item(DataCategory::LogByte), self.bytes())
            .await;

        let limits = items.merge_with(bytes);
        if !limits.is_empty() {
            return Err(self.reject_err(Error::RateLimited(limits)));
        }

        Ok(())
    }
}
