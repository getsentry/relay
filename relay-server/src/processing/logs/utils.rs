use relay_event_schema::protocol::OurLog;

use crate::envelope::WithHeader;

/// Returns the calculated size of a [`OurLog`].
///
/// Unlike [`relay_ourlogs::calculate_size`], this accesses the already manifested byte size
/// of the log, instead of calculating it.
///
/// When compiled with debug assertion the function asserts the presence of a manifested byte size.
pub fn get_calculated_byte_size(log: &WithHeader<OurLog>) -> usize {
    // Use the remembered byte size here, as this is the one Relay stored when initially
    // receiving the log item. This will not include any modifications done to the log.
    let bytes = log.header.as_ref().and_then(|header| header.byte_size);

    debug_assert!(
        bytes.is_some(),
        "processed logs should always have a byte size assigned"
    );

    // Only fall back to a calculated size, when there is nothing stored.
    // This should never happen, logs should have a size assigned to them immediately after
    // they have been expanded.
    let bytes = bytes.unwrap_or_else(|| log.value().map_or(1, relay_ourlogs::calculate_size));

    usize::try_from(bytes).unwrap_or(usize::MAX)
}
