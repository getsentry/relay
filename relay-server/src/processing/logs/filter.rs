use relay_dynamic_config::Feature;
use relay_event_schema::protocol::OurLog;
use relay_protocol::Annotated;

use crate::extractors::RequestMeta;
use crate::managed::Managed;
use crate::processing::Context;
use crate::processing::logs::{Error, ExpandedLogs, Result};

/// Filters logs sent for a project which does not allow logs ingestion.
pub fn feature_flag(ctx: Context<'_>) -> Result<()> {
    match ctx.should_filter(Feature::OurLogsIngestion) {
        true => Err(Error::FilterFeatureFlag),
        false => Ok(()),
    }
}

/// Applies inbound filters to individual logs.
pub fn filter(logs: &mut Managed<ExpandedLogs>, ctx: Context<'_>) {
    logs.retain_with_context(
        |logs| (&mut logs.logs, logs.headers.meta()),
        |log, meta, _| filter_log(log, meta, ctx),
    );
}

fn filter_log(log: &Annotated<OurLog>, meta: &RequestMeta, ctx: Context<'_>) -> Result<()> {
    let Some(log) = log.value() else {
        return Ok(());
    };

    relay_filter::should_filter(
        log,
        meta.client_addr(),
        &ctx.project_info.config.filter_settings,
        ctx.global_config.filters(),
    )
    .map_err(Error::Filtered)
}
