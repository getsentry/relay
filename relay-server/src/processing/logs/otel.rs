use prost::Message as _;
use relay_event_schema::protocol::{OurLog, OurLogHeader};
use relay_ourlogs::otel_logs::LogsData;
use relay_quotas::DataCategory;

use crate::envelope::{ContainerItems, ContentType, Item, ItemType, WithHeader};
use crate::managed::RecordKeeper;
use crate::processing::logs::Error;
use crate::services::outcome::DiscardReason;

pub fn expand_into(
    result: &mut ContainerItems<OurLog>,
    records: &mut RecordKeeper<'_>,
    items: Vec<Item>,
) {
    for item in items {
        debug_assert_eq!(item.ty(), &ItemType::OtelLogsData);

        let logs = match parse_logs_data(&item) {
            Ok(logs) => logs,
            Err(err) => {
                records.reject_err(err, &item);
                continue;
            }
        };

        // Undo all the base item quantities, as they will be completely taken over by the parsed
        // contents, which contains an arbitrary amount of items (even 0).
        for (category, quantity) in item.quantities() {
            records.modify_by(category, -(quantity as isize));
        }

        for resource_logs in logs.resource_logs {
            let resource = resource_logs.resource.as_ref();
            for scope_logs in resource_logs.scope_logs {
                let scope = scope_logs.scope.as_ref();
                for log_record in scope_logs.log_records {
                    let log = relay_ourlogs::otel_to_sentry_log(log_record, resource, scope);
                    let byte_size = relay_ourlogs::calculate_size(&log);

                    records.modify_by(DataCategory::LogItem, 1);
                    records.modify_by(DataCategory::LogByte, byte_size as isize);

                    result.push(WithHeader {
                        header: Some(OurLogHeader {
                            byte_size: Some(byte_size),
                            other: Default::default(),
                        }),
                        value: log.into(),
                    });
                }
            }
        }
    }
}

fn parse_logs_data(item: &Item) -> Result<LogsData, Error> {
    match item.content_type() {
        Some(&ContentType::Json) => serde_json::from_slice(&item.payload()).map_err(|e| {
            relay_log::debug!(
                error = &e as &dyn std::error::Error,
                "Failed to parse logs data as JSON"
            );
            Error::Invalid(DiscardReason::InvalidJson)
        }),
        Some(&ContentType::Protobuf) => LogsData::decode(item.payload()).map_err(|e| {
            relay_log::debug!(
                error = &e as &dyn std::error::Error,
                "Failed to parse logs data as protobuf"
            );
            Error::Invalid(DiscardReason::InvalidProtobuf)
        }),
        _ => Err(Error::Invalid(DiscardReason::ContentType)),
    }
}
