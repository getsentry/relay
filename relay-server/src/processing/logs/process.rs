use chrono::{DateTime, Utc};
use relay_event_normalization::{
    ClientHints, FromUserAgentInfo as _, RawUserAgentInfo, SchemaProcessor,
};
use relay_event_schema::processor::{ProcessingState, process_value};
use relay_event_schema::protocol::{AttributeType, BrowserContext, OurLog, OurLogHeader};
use relay_ourlogs::OtelLog;
use relay_pii::PiiProcessor;
use relay_protocol::{Annotated, ErrorKind, Value};
use relay_quotas::DataCategory;

use crate::envelope::{ContainerItems, Item, ItemContainer, WithHeader};
use crate::extractors::{RequestMeta, RequestTrust};
use crate::processing::logs::{Error, ExpandedLogs, Result, SerializedLogs};
use crate::processing::{Context, Managed};
use crate::services::outcome::DiscardReason;

/// Parses all serialized logs into their [`ExpandedLogs`] representation.
///
/// Individual, invalid logs will be discarded.
pub fn expand(logs: Managed<SerializedLogs>, _ctx: Context<'_>) -> Managed<ExpandedLogs> {
    let received_at = logs.received_at();
    let trust = logs.headers.meta().request_trust();

    logs.map(|logs, records| {
        records.lenient(DataCategory::LogByte);

        let mut all_logs = Vec::with_capacity(logs.count());
        for logs in logs.logs {
            let expanded = expand_log_container(&logs, received_at, trust);
            let expanded = records.or_default(expanded, logs);
            all_logs.extend(expanded);
        }

        for otel_log in logs.otel_logs {
            match expand_otel_log(&otel_log, received_at) {
                Ok(log) => all_logs.push(log),
                Err(err) => {
                    records.reject_err(err, otel_log);
                    continue;
                }
            }
        }

        ExpandedLogs {
            headers: logs.headers,
            #[cfg(feature = "processing")]
            retention: _ctx.project_info.config.event_retention,
            #[cfg(feature = "processing")]
            serialize_meta_attrs: _ctx
                .project_info
                .has_feature(relay_dynamic_config::Feature::OurLogsMetaAttributes),
            logs: all_logs,
        }
    })
}

/// Processes expanded logs.
///
/// Validates, scrubs, normalizes and enriches individual log entries.
pub fn process(logs: &mut Managed<ExpandedLogs>, ctx: Context<'_>) {
    logs.modify(|logs, records| {
        let meta = logs.headers.meta();
        logs.logs
            .retain_mut(|log| records.or_default(process_log(log, meta, ctx).map(|_| true), &*log));
    });
}

fn expand_otel_log(item: &Item, received_at: DateTime<Utc>) -> Result<WithHeader<OurLog>> {
    let log = serde_json::from_slice::<OtelLog>(&item.payload()).map_err(|err| {
        relay_log::debug!("failed to parse OTel Log: {err}");
        Error::Invalid(DiscardReason::InvalidJson)
    })?;

    let log = relay_ourlogs::otel_to_sentry_log(log, received_at).map_err(|err| {
        relay_log::debug!("failed to convert OTel Log to Sentry Log: {:?}", err);
        Error::Invalid(DiscardReason::InvalidLog)
    })?;

    // The OTel log conversion already adds certain Sentry attributes which are included in the
    // cost here.
    //
    // As OTel logs are deprecated and to be removed, this is okay for now.
    //
    // See: <https://github.com/getsentry/relay/issues/4884>.
    let byte_size = Some(relay_ourlogs::calculate_size(&log));
    Ok(WithHeader {
        value: Annotated::new(log),
        header: Some(OurLogHeader {
            byte_size,
            other: Default::default(),
        }),
    })
}

fn expand_log_container(
    item: &Item,
    received_at: DateTime<Utc>,
    trust: RequestTrust,
) -> Result<ContainerItems<OurLog>> {
    let mut logs = ItemContainer::parse(item)
        .map_err(|err| {
            relay_log::debug!("failed to parse logs container: {err}");
            Error::Invalid(DiscardReason::InvalidJson)
        })?
        .into_items();

    for log in &mut logs {
        // Calculate the received byte size and remember it as metadata, in the header.
        // As Relay may later modify the payload later, adding attributes or removing them
        // and in any case we want to track the originally received size.
        //
        // Since Relay can be deployed in multiple layers, we need to remember the size of the
        // first Relay which received the log, but at the same time we must be able to trust that
        // size.
        //
        // If there is no size calculated or we cannot trust the source, we re-calculate the size.
        let byte_size = log.header.as_ref().and_then(|h: &OurLogHeader| h.byte_size);
        if trust.is_untrusted() || matches!(byte_size, None | Some(0)) {
            let byte_size = log.value().map(relay_ourlogs::calculate_size).unwrap_or(1);
            log.header.get_or_insert_default().byte_size = Some(byte_size);
        }

        relay_ourlogs::ourlog_merge_otel(log, received_at);
    }

    Ok(logs)
}

fn process_log(log: &mut Annotated<OurLog>, meta: &RequestMeta, ctx: Context<'_>) -> Result<()> {
    scrub(log, ctx).inspect_err(|err| {
        relay_log::debug!("failed to scrub pii from log: {err}");
    })?;

    normalize(log, meta).inspect_err(|err| {
        relay_log::debug!("failed to normalize log: {err}");
    })?;

    Ok(())
}

fn scrub(log: &mut Annotated<OurLog>, ctx: Context<'_>) -> Result<()> {
    let pii_config = ctx
        .project_info
        .config
        .datascrubbing_settings
        .pii_config()
        .map_err(|e| Error::PiiConfig(e.clone()))?;

    if let Some(ref config) = ctx.project_info.config.pii_config {
        let mut processor = PiiProcessor::new(config.compiled());
        process_value(log, &mut processor, ProcessingState::root())?;
    }

    if let Some(config) = pii_config {
        let mut processor = PiiProcessor::new(config.compiled());
        process_value(log, &mut processor, ProcessingState::root())?;
    }

    Ok(())
}

fn normalize(log: &mut Annotated<OurLog>, meta: &RequestMeta) -> Result<()> {
    process_value(log, &mut SchemaProcessor, ProcessingState::root())?;

    let Some(log) = log.value_mut() else {
        return Err(Error::Invalid(DiscardReason::NoData));
    };

    process_attribute_types(log);
    populate_ua_fields(log, meta.user_agent(), meta.client_hints().as_deref());

    Ok(())
}

fn populate_ua_fields(log: &mut OurLog, user_agent: Option<&str>, client_hints: ClientHints<&str>) {
    let attributes = log.attributes.get_or_insert_with(Default::default);

    const BROWSER_NAME: &str = "sentry.browser.name";
    const BROWSER_VERSION: &str = "sentry.browser.version";

    if attributes.contains_key(BROWSER_NAME) && attributes.contains_key(BROWSER_VERSION) {
        return;
    }

    let Some(context) = BrowserContext::from_hints_or_ua(&RawUserAgentInfo {
        user_agent,
        client_hints,
    }) else {
        return;
    };

    if let Some(name) = context.name.into_value() {
        attributes.insert_if_missing(BROWSER_NAME, || name);
    }
    if let Some(version) = context.version.into_value() {
        attributes.insert_if_missing(BROWSER_VERSION, || version);
    }
}

fn process_attribute_types(ourlog: &mut OurLog) {
    let Some(attributes) = ourlog.attributes.value_mut() else {
        return;
    };

    let attributes = attributes.iter_mut().map(|(_, attr)| attr);

    for attribute in attributes {
        use AttributeType::*;

        let Some(inner) = attribute.value_mut() else {
            continue;
        };

        match (&mut inner.value.ty, &mut inner.value.value) {
            (Annotated(Some(Boolean), _), Annotated(Some(Value::Bool(_)), _)) => (),
            (Annotated(Some(Integer), _), Annotated(Some(Value::I64(_)), _)) => (),
            (Annotated(Some(Integer), _), Annotated(Some(Value::U64(_)), _)) => (),
            (Annotated(Some(Double), _), Annotated(Some(Value::I64(_)), _)) => (),
            (Annotated(Some(Double), _), Annotated(Some(Value::U64(_)), _)) => (),
            (Annotated(Some(Double), _), Annotated(Some(Value::F64(_)), _)) => (),
            (Annotated(Some(String), _), Annotated(Some(Value::String(_)), _)) => (),
            // Note: currently the mapping to Kafka requires that invalid or unknown combinations
            // of types and values are removed from the mapping.
            //
            // Usually Relay would only modify the offending values, but for now, until there
            // is better support in the pipeline here, we need to remove the entire attribute.
            (Annotated(Some(Unknown(_)), _), _) => {
                let original = attribute.value_mut().take();
                attribute.meta_mut().add_error(ErrorKind::InvalidData);
                attribute.meta_mut().set_original_value(original);
            }
            (Annotated(Some(_), _), Annotated(Some(_), _)) => {
                let original = attribute.value_mut().take();
                attribute.meta_mut().add_error(ErrorKind::InvalidData);
                attribute.meta_mut().set_original_value(original);
            }
            (Annotated(None, _), _) | (_, Annotated(None, _)) => {
                let original = attribute.value_mut().take();
                attribute.meta_mut().add_error(ErrorKind::MissingAttribute);
                attribute.meta_mut().set_original_value(original);
            }
        }
    }
}

#[cfg(test)]
mod tests {
    use relay_protocol::SerializableAnnotated;

    use super::*;

    #[test]
    fn test_process_attribute_types() {
        let json = r#"{
            "timestamp": 1544719860.0,
            "trace_id": "5b8efff798038103d269b633813fc60c",
            "span_id": "eee19b7ec3c1b174",
            "level": "info",
            "body": "Test log message",
            "attributes": {
                "valid_bool": {
                    "type": "boolean",
                    "value": true
                },
                "valid_int_i64": {
                    "type": "integer",
                    "value": -42
                },
                "valid_int_u64": {
                    "type": "integer",
                    "value": 42
                },
                "valid_int_from_string": {
                    "type": "integer",
                    "value": "42"
                },
                "valid_double": {
                    "type": "double",
                    "value": 42.5
                },
                "double_with_i64": {
                    "type": "double",
                    "value": -42
                },
                "valid_double_with_u64": {
                    "type": "double",
                    "value": 42
                },
                "valid_string": {
                    "type": "string",
                    "value": "test"
                },
                "valid_string_with_other": {
                    "type": "string",
                    "value": "test",
                    "some_other_field": "some_other_value"
                },
                "unknown_type": {
                    "type": "custom",
                    "value": "test"
                },
                "invalid_int_from_invalid_string": {
                    "type": "integer",
                    "value": "abc"
                },
                "missing_type": {
                    "value": "value with missing type"
                },
                "missing_value": {
                    "type": "string"
                }
            }
        }"#;

        let mut data = Annotated::<OurLog>::from_json(json).unwrap();

        if let Some(log) = data.value_mut() {
            process_attribute_types(log);
        }

        insta::assert_json_snapshot!(SerializableAnnotated(&data.value().unwrap().attributes), @r###"
        {
          "double_with_i64": {
            "type": "double",
            "value": -42
          },
          "invalid_int_from_invalid_string": null,
          "missing_type": null,
          "missing_value": null,
          "unknown_type": null,
          "valid_bool": {
            "type": "boolean",
            "value": true
          },
          "valid_double": {
            "type": "double",
            "value": 42.5
          },
          "valid_double_with_u64": {
            "type": "double",
            "value": 42
          },
          "valid_int_from_string": null,
          "valid_int_i64": {
            "type": "integer",
            "value": -42
          },
          "valid_int_u64": {
            "type": "integer",
            "value": 42
          },
          "valid_string": {
            "type": "string",
            "value": "test"
          },
          "valid_string_with_other": {
            "type": "string",
            "value": "test",
            "some_other_field": "some_other_value"
          },
          "_meta": {
            "invalid_int_from_invalid_string": {
              "": {
                "err": [
                  "invalid_data"
                ],
                "val": {
                  "type": "integer",
                  "value": "abc"
                }
              }
            },
            "missing_type": {
              "": {
                "err": [
                  "missing_attribute"
                ],
                "val": {
                  "type": null,
                  "value": "value with missing type"
                }
              }
            },
            "missing_value": {
              "": {
                "err": [
                  "missing_attribute"
                ],
                "val": {
                  "type": "string",
                  "value": null
                }
              }
            },
            "unknown_type": {
              "": {
                "err": [
                  "invalid_data"
                ],
                "val": {
                  "type": "custom",
                  "value": "test"
                }
              }
            },
            "valid_int_from_string": {
              "": {
                "err": [
                  "invalid_data"
                ],
                "val": {
                  "type": "integer",
                  "value": "42"
                }
              }
            }
          }
        }
        "###);
    }

    #[test]
    fn test_populate_ua_fields() {
        let mut log = OurLog::default();
        populate_ua_fields(
            &mut log,
            Some(
                "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/131.0.0.0 Safari/537.36",
            ),
            ClientHints::default(),
        );

        let attributes = log.attributes.value().unwrap();
        assert_eq!(
            attributes.get_value("sentry.browser.name").unwrap(),
            &Value::String("Chrome".to_owned()),
        );
        assert_eq!(
            attributes.get_value("sentry.browser.version").unwrap(),
            &Value::String("131.0.0".to_owned()),
        );
    }
}
