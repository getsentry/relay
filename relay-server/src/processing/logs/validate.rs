use crate::processing::Managed;
use crate::processing::logs::{Error, Result, SerializedLogs};

/// Validates that there is only a single log container processed at a time.
///
/// The `Log` item must always be sent as an `ItemContainer`, currently it is not allowed to
/// send multiple containers for logs.
///
/// This restriction may be lifted in the future.
///
/// This limit mostly exists to incentivise SDKs to batch multiple logs into a single container,
/// technically it can be removed without issues.
pub fn container(logs: &Managed<SerializedLogs>) -> Result<()> {
    // It's fine if there was no log container, as we still accept OTel logs.
    if logs.logs.len() > 1 {
        return Err(Error::DuplicateContainer);
    }

    Ok(())
}
