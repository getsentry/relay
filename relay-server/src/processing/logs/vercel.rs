use relay_event_schema::protocol::{OurLog, OurLogHeader, VercelLog};
use relay_ourlogs::vercel_to_sentry_log;
use relay_quotas::DataCategory;

use crate::envelope::{ContainerItems, ContentType, Item, ItemType, WithHeader};
use crate::managed::RecordKeeper;
use crate::processing::logs::Error;
use crate::services::outcome::DiscardReason;

pub fn expand_into(
    result: &mut ContainerItems<OurLog>,
    records: &mut RecordKeeper<'_>,
    items: Vec<Item>,
) {
    for item in items {
        debug_assert_eq!(item.ty(), &ItemType::VercelLog);

        let vercel_logs = match parse_vercel_logs(&item) {
            Ok(logs) => logs,
            Err(err) => {
                records.reject_err(err, &item);
                continue;
            }
        };

        // Undo all the base item quantities, as they will be completely taken over by the parsed
        // contents, which contains an arbitrary amount of items (even 0).
        for (category, quantity) in item.quantities() {
            records.modify_by(category, -(quantity as isize));
        }

        for vercel_log in vercel_logs {
            let log = vercel_to_sentry_log(vercel_log);
            let byte_size = relay_ourlogs::calculate_size(&log);

            records.modify_by(DataCategory::LogItem, 1);
            records.modify_by(DataCategory::LogByte, byte_size as isize);

            result.push(WithHeader {
                header: Some(OurLogHeader {
                    byte_size: Some(byte_size),
                    other: Default::default(),
                }),
                value: log.into(),
            });
        }
    }
}

fn parse_vercel_logs(item: &Item) -> Result<Vec<VercelLog>, Error> {
    match item.content_type() {
        Some(&ContentType::Json) => {
            // Vercel logs are expected to be sent as a JSON array
            serde_json::from_slice::<Vec<VercelLog>>(&item.payload()).map_err(|e| {
                relay_log::debug!(
                    error = &e as &dyn std::error::Error,
                    "Failed to parse Vercel logs as JSON array"
                );
                Error::Invalid(DiscardReason::InvalidJson)
            })
        }
        _ => Err(Error::Invalid(DiscardReason::ContentType)),
    }
}
