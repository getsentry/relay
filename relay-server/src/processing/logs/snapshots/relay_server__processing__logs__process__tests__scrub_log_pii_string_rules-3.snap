---
source: relay-server/src/processing/logs/process.rs
expression: SerializableAnnotated(&data.value().unwrap().attributes)
---
{
  "@ip:remove|127.0.0.1": {
    "type": "string",
    "value": ""
  },
  "_meta": {
    "@ip:remove|127.0.0.1": {
      "value": {
        "": {
          "rem": [
            [
              "@ip:remove",
              "x",
              0,
              0
            ]
          ],
          "len": 9
        }
      }
    }
  }
}
